include/pybind11/gil_safe_call_once.h

```diff
@@ -0,0 +1,85 @@
+//
+//   static auto imported_obj = py::module_::import("module_name"); // BAD, DO NOT USE!
+//
+// which has two serious issues:
+//
+//     1. Py_DECREF() calls potentially after the Python interpreter was finalized already, and
+//     2. deadlocks in multi-threaded processes (because of missing lock ordering).
+//
+// The following alternative avoids both problems:
+//
+//   PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object> storage;
+//   auto &imported_obj = storage // Do NOT make this `static`!
+//       .call_once_and_store_result([]() {
+//           return py::module_::import("module_name");
+//       })
+//       .get_stored();
+//
+// The parameter of `call_once_and_store_result()` must be callable. It can make
+// CPython API calls, and in particular, it can temporarily release the GIL.
+//
+// `T` can be any C++ type, it does not have to involve CPython API types.
+//
+// The behavior with regard to signals, e.g. `SIGINT` (`KeyboardInterrupt`),
+// is not ideal. If the main thread is the one to actually run the `Callable`,
+// then a `KeyboardInterrupt` will interrupt it if it is running normal Python
+// code. The situation is different if a non-main thread runs the
+// `Callable`, and then the main thread starts waiting for it to complete:
+// a `KeyboardInterrupt` will not interrupt the non-main thread, but it will
+// get processed only when it is the main thread's turn again and it is running
+// normal Python code. However, this will be unnoticeable for quick call-once
+// functions, which is usually the case.
+template <typename T>
+class gil_safe_call_once_and_store {
+public:
+    // PRECONDITION: The GIL must be held when `call_once_and_store_result()` is called.
+    template <typename Callable>
+    gil_safe_call_once_and_store &call_once_and_store_result(Callable &&fn) {
+        if (!is_initialized_.load(std::memory_order_acquire)) {
+            gil_scoped_release gil_rel;
+            std::call_once(once_flag_, [&] {
+                gil_scoped_acquire gil_acq;
+                ::new (storage_) T(fn());
+                is_initialized_.store(true, std::memory_order_release);
+            });
+        }
+        return *this;
+    }
+
+    // This must only be called after `call_once_and_store()` was called.
+    // Not const for simplicity. (Could be made const if there is an unforeseen need.)
+    T &get_stored() {
+        assert(is_initialized_.load(std::memory_order_relaxed));
+        PYBIND11_WARNING_PUSH
+#if !defined(__clang__) && defined(__GNUC__) && __GNUC__ < 5
+        // Needed for gcc 4.8.5
+        PYBIND11_WARNING_DISABLE_GCC("-Wstrict-aliasing")
+#endif
+        return *reinterpret_cast<T *>(storage_);
+        PYBIND11_WARNING_POP
+    }
+
+    constexpr gil_safe_call_once_and_store() = default;
+    PYBIND11_DTOR_CONSTEXPR ~gil_safe_call_once_and_store() = default;
+
+private:
+    alignas(T) char storage_[sizeof(T)] = {};
```

Comment:
I would make this a std::optional<T> and avoid both the atomic  flag and the storage here. 

You don't need the atomic<bool> since call_once is an inherent mutex that asserts ordering.

Reply:
It's not the access inside the `call_once` that the atomic is for, but rather, the earler access (the first check for the fast path).

However, I think you have a point that because we assume that the GIL is held on function entry, the boolean doesn't need to be atomic: the GIL serializes the outer check, and the once-mutex serializes the inner check. The code with the atomic variable would work even without any external synchronisation, but since we already have the GIL on entry, I agree that we should be able to use a non-atomic boolean.

(And then the boolean + placement new could be replaced by std::optional, indeed.)

Reply:
And just to make this point clearly: the first, outer boolean check provides a fast path on which we don't interact with the GIL any further. We only do the release and reacquire dance on the (hopefully rare) slow path.

Reply:
Thanks Thomas! I'll definitely make the change to the simple `bool` then.

> could be replaced by std::optional

Except that pybind11 still supports C++11, and `std::optional` was added only with C++17.

Reply:
In that case leave it as bool + placement new :-)

Reply:
Ah, and also we do actually want the "constant-initializable, trivial destructible" behaviour of bool + placement new. Not only do we not want to destroy this object, but this also gives us cheaper static variables (no static guard, and no global destructructor list entry).